modules/foreman_facts.py

```diff
@@ -0,0 +1,157 @@
+#!/usr/bin/python
+# -*- coding: utf-8 -*-
+# (c) 2018, Sean O'Keeffe <seanokeeffe797@gmail.com>
+#
+# This file is part of Ansible
+#
+# Ansible is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# Ansible is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with Ansible.  If not, see <http://www.gnu.org/licenses/>.
+
+ANSIBLE_METADATA = {'metadata_version': '1.0',
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: foreman_facts
+short_description: Gather facts about Foreman Entities
+description:
+  - "Gather facts about Foreman Entites"
+author:
+  - "Sean O'Keeffe (@sean797)"
+requirements:
+  - nailgun
+options:
+  server_url:
+    description:
+      - URL of Foreman server
+    required: true
+  username:
+    description:
+      - Username on Foreman server
+    required: true
+  password:
+    description:
+      - Password for user accessing Foreman server
+    required: true
+  verify_ssl:
+    description:
+      - Verify SSL of the Foreman server
+    default: true
+    type: bool
+  entity:
+    description:
+      - Entity to search
+      - Run with paramater set to C(foo) to see all available options.
+  search:
+    description:
+      - Search query to use
+      - If None, all entities are returned
+'''
+
+EXAMPLES = '''
+- name: "Read a Setting"
+  foreman_facts:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    entity = Setting
+    search: name = http_proxy
+  register: result
+- debug:
+    var: result.response[0].value
+
+- name: "Read all Registries"
+  foreman_facts:
+    username: "admin"
+    password: "changeme"
+    server_url: "https://foreman.example.com"
+    entity = Registry
+  register: result
+- debug:
+    var: item.name
+  with_items: result.response
+'''
+
+RETURN = '''
+response:
+  description: Data returned from Foreman
+'''
+
+try:
+    from ansible.module_utils.ansible_nailgun_cement import (
+        create_server,
+        ping_server,
+        search_entities_json,
+    )
+
+    from nailgun import entities
+
+    has_import_error = False
+except ImportError as e:
+    has_import_error = True
+    import_error_msg = str(e)
+    raise
+
+from ansible.module_utils.basic import AnsibleModule
+
+
+def nailgun_entites():
+    classes_to_strip = ['APIResponseError', 'Entity', 'EntityCreateMixin', 'EntityDeleteMixin', 'EntityReadMixin', 'EntitySearchMixin', 'EntityUpdateMixin']
```

Comment:
This looks like a blacklist. Wouldn't it be better to filter classes by inheritance of `EntitySearchMixin` and `EntityReadMixin`. I think, that would give you exactly what we need here.